### ISSUE TITLE
Slurm `gpus-per-task` Error

### ISSUE THREAD

**Issue by Queuecumber** [Lightning-AI/lightning#16828]
### Bug description

We recently uncovered what I believe is a PL but in the way slurm multi-gpu is being handled. This seems to be caused by a change in slurm behavior for some recent versions (we haven't nailed down exactly when this change happened)

The error is caused by using the `--gpus-per-task` slurm argument (potentially other GPU arguments as well). For example if I have a machine with 8 GPUs I will say `--gpus-per-task=1` and `--tasks=8` to launch 8 processes with 1 GPU each. Previous versions of slurm didnt seem to enforce GPU isolation but the newer versions are setting `CUDA_VISIBLE_DEVICES` appropriately. 

The problem is that lightning seems to expect to see all GPUs on a node from all processes, so it errors out with 

```
MisconfigurationException('You requested gpu: [0, 1, 2, 3, 4, 5, 6, 7]\n But your machine only has: [0]
```

In this case, I think slurm is doing the right thing since this would prevent tasks from accidentally using GPUs that they werent assigned. I also think that PL shouldn't care if it can see multiple GPUs or not as long as all distributed processes check in with the primary. Since slurm is handling the multiprocessing, and PL knows this, it should only need to see a single GPU in order to work.

Is this a bug that should be fixed?

### How to reproduce the bug

```python
Here's a redacted sbatch that reproduces


#!/bin/bash

# Parameters
#SBATCH --cpus-per-gpu=12
#SBATCH --gpus-per-task=1
#SBATCH --mem-per-gpu=112500
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=8
#SBATCH --open-mode=append
#SBATCH --requeue
#SBATCH --signal=USR2@120
#SBATCH --time=240
```

The lightning command should be set up for 8 GPUs on 1 node.
```


### Error messages and logs

```
MisconfigurationException('You requested gpu: [0, 1, 2, 3, 4, 5, 6, 7]\n But your machine only has: [0]
```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>


### More info

_No response_

**Comment by Queuecumber** [comment]
Yeah that sounds perfect, so `devices=auto` already handles this case? Does it also work with multi node?

**Comment by awaelchli** [comment]
`devices=auto` doesn't handle this case for slurm (yet), I just think it would fit well together with your request here and motivate it even more so. 

One question, in your script above, is the setting `SBATCH --gpus-per-task=1`, if you leave it away, what will happen? Does the newer version of slurm still restrict it to 1 visible gpu per process or not?

**Comment by Queuecumber** [comment]
If I don't pass a GPU argument (e.g., omitting `gpus-per-task`) it won't make any GPUs visible to the task. In other words the task will think there are 0 GPUs on the machine.

What we did you fix it is it set a node-level control on the GPUs, specifically we passed `--gres:gpu=8` which makes sure that the node itself has 8 GPUs available, then any tasks can see all 8 GPUs. It's actually pretty confusing how all these things play off each other.

Do you need a PR for `gpus=auto` slurm support or is that something that still needs discussion/other supporting code?

**Comment by rndompc** [comment]
> If I don't pass a GPU argument (e.g., omitting `gpus-per-task`) it won't make any GPUs visible to the task. In other words the task will think there are 0 GPUs on the machine.
> 
> What we did to fix it is it set a node-level control on the GPUs, specifically we passed `--gres:gpu=8` which makes sure that the node itself has 8 GPUs available, then any tasks can see all 8 GPUs. It's actually pretty confusing how all these things play off each other.
> 
> Do you need a PR for `gpus=auto` slurm support or is that something that still needs discussion/other supporting code?

Did you make it work? if so how does you slurm file looks now? I am having the same issue  in my case it prints:

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
Rank: 0 partition count [2, 2] and sizes[(757553152, False), (768, False)]
Rank: 1 partition count [2, 2] and sizes[(757553152, False), (768, False)]


but only uses 1 gpu I am using 2 nodes and in each node there are 8 gpus.

**Comment by Queuecumber** [comment]
The only thing I changed was to replace `--gpus-per-task=1` with `--gres:gpu=8` this seems to  allow all tasks to view all 8 GPUs which prevents lightning from getting confused about how many GPUs are available on the node

**Comment by alexanderwerning** [comment]
I encountered a similar issue with `--ntasks-per-node`. for `devices=4`, pytorch-lightning complains that there are not enough gpus (only one is visible per task), for `devices=1`, the mismatch between `devices` and `--ntasks-per-node` is a problem.
The workaround with `--gres:gpu=4` does not work, only one gpu is visible to each task still...

**Comment by EnsiyehRaoufi** [comment]
> I encountered a similar issue with `--ntasks-per-node`. for `devices=4`, pytorch-lightning complains that there are not enough gpus (only one is visible per task), for `devices=1`, the mismatch between `devices` and `--ntasks-per-node` is a problem. The workaround with `--gres:gpu=4` does not work, only one gpu is visible to each task...

I have the same problem. Did you solve the issue? If yes, can you explain to me how? Thanks

**Comment by alexanderwerning** [comment]
Yes, I solved it by adding `export CUDA_VISIBLE_DEVICES=0,1,2,3` to the slurm script, explicitly making all gpus visible to every process. I would prefer a solution where each process only sees a single gpu, but this does not seem to be possible.

**Comment by EnsiyehRaoufi** [comment]
> Yes, I solved it by adding `export CUDA_VISIBLE_DEVICES=0,1,2,3` to the slurm script, explicitly making all gpus visible to every process. This is necessary since slurm automatically sets unique CUDA_VISIBLE_DEVICES per task, see https://slurm.schedmd.com/gres.html#GPU_Management I would prefer a solution where each process only sees a single gpu, but this does not seem to be possible.

Thank you so much for your quick response. Unfortunately, your solution did not work for me. 
I wanted to load the model using an `AutoModel.from_pretrained(model_name, device_map="auto")` module, and while I've reserved 8 GPUs per node, I was facing "torch.cuda.OutOfMemoryError". Inspiring from [here](https://github.com/Lightning-AI/pytorch-lightning/issues/18950), I could successfully load the model on several GPUs by changing my Slurm config to the following: `--nodes=1` (could be set to a larger number), `--ntasks-per-node=8`, `--gpus=4`. 
However, when I print the number of GPUs per node using environmental variables in my Python code, it shows that 8 GPUs are available. So, I don't know what the number 4 in `--gpus=4` exactly indicates (when I remove it from Slurm config, I receive this error: "Job type CPU. sbatch: error: Batch job submission failed: Invalid account or account/partition combination specific").
 
I appreciate it if someone guides me.

**Comment by Elias-Buerger** [comment]
Has there been any progress on this? I am struggling with this and no workaround seems to be effective
